include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -25,295 +25,105 @@ namespace nbl::asset
 		{
-			for (uint32_t t = 0u; t < EDT_COUNT; ++t)
-			{
-				if (m_descriptors[t])
-					m_descriptorInfos[t] = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUDescriptorSet::SDescriptorInfo::SBufferImageInfo>>(m_descriptors[t]->size());
-			}
-		}
+			const auto type = static_cast<IDescriptor::E_TYPE>(t);
+			const uint32_t count = m_layout->getTotalDescriptorCount(type);
+			if (count == 0u)
+				continue;
 
-		inline size_t conservativeSizeEstimate() const override
-		{
-			assert(!"Invalid code path.");
-			return 0xdeadbeefull;
+			m_descriptorInfos[t] = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUDescriptorSet::SDescriptorInfo>>(count);
 		}
+	}
 
-        core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const override
-        {
-            auto layout = (_depth > 0u && m_layout) ? core::smart_refctd_ptr_static_cast<ICPUDescriptorSetLayout>(m_layout->clone(_depth - 1u)) : m_layout;
-            auto cp = core::make_smart_refctd_ptr<ICPUDescriptorSet>(std::move(layout));
-            clone_common(cp.get());
+	_NBL_STATIC_INLINE_CONSTEXPR auto AssetType = ET_DESCRIPTOR_SET;
+	inline E_TYPE getAssetType() const override { return AssetType; }
 
-			auto cloneDescriptor = [](const core::smart_refctd_ptr<IDescriptor>& _desc, uint32_t _depth) -> core::smart_refctd_ptr<IDescriptor>
-			{
-				if (!_desc)
-					return nullptr;
+	inline ICPUDescriptorSetLayout* getLayout() 
+	{
+		assert(!isImmutable_debug());
+		return m_layout.get();
+	}
 
-				IAsset* asset = nullptr;
-				switch (_desc->getTypeCategory())
-				{
-				case IDescriptor::EC_BUFFER:
-					asset = static_cast<ICPUBuffer*>(_desc.get()); break;
-				case IDescriptor::EC_BUFFER_VIEW:
-					asset = static_cast<ICPUBufferView*>(_desc.get()); break;
-				case IDescriptor::EC_IMAGE:
-					asset = static_cast<ICPUImageView*>(_desc.get()); break;
-				}
+	inline const ICPUDescriptorSetLayout* getLayout() const { return m_layout.get(); }
 
-				auto cp = asset->clone(_depth);
+	inline bool canBeRestoredFrom(const IAsset* _other) const override
+	{
+		auto* other = static_cast<const ICPUDescriptorSet*>(_other);
+		return m_layout->canBeRestoredFrom(other->m_layout.get());
+	}
 
-				switch (_desc->getTypeCategory())
-				{
-				case IDescriptor::EC_BUFFER:
-					return core::smart_refctd_ptr_static_cast<ICPUBuffer>(std::move(cp));
-				case IDescriptor::EC_BUFFER_VIEW:
-					return core::smart_refctd_ptr_static_cast<ICPUBufferView>(std::move(cp));
-				case IDescriptor::EC_IMAGE:
-					return core::smart_refctd_ptr_static_cast<ICPUImageView>(std::move(cp));
-				}
-				return nullptr;
-			};
+	inline size_t conservativeSizeEstimate() const override
+	{
+		assert(!"Invalid code path.");
+		return 0xdeadbeefull;
+	}
 
-			for (uint32_t t = 0u; t < EDT_COUNT; ++t)
-			{
-				const auto type = static_cast<E_DESCRIPTOR_TYPE>(t);
+	inline core::SRange<SDescriptorInfo> getDescriptorInfoStorage(const IDescriptor::E_TYPE type) const
```

Comment:
you're not returning a `core::SRange<const SDescriptorInfo>` so a mutability check is in order

P.S. see some of the functions in the engine where we use `const_cast` to quickly implement both `const` non-`const` overloads

Reply:
Same as this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1054258384

First we need to fix the usages in the asset converter and think about its cascading effects.. or use `const_cast` somewhere along the call chain.

Reply:
Really, the real culprit is [this patch-up work we do in the asset converter](https://github.com/Devsh-Graphics-Programming/Nabla/blob/22284fd82eb56e2dcec09df4f486d620bb82b474/include/nbl/video/utilities/IGPUObjectFromAssetConverter.h#L1644-L1648). Probably we should just use a `const_cast` in the asset converter and make a note to fix it during the development of 2.0.

Reply:
@achalpandeyy put the mutability checks there, comment them out and put a link to these convos in the comments

Reply:
@Hazardu this is the reason for the mapping of Asset to CPU Object + Metadata so we can patch up the usage flags and shit while keeping the Assets const

Reply:
Resolved.